### DISCUSSION TITLE
Relative path in `environments.*.values` not working in multi-file setup.

### DISCUSSION THREAD

**Discussion by Kellen275** [helmfile/helmfile#1301] [Questions]
Given the following file structure
```shell
.
├── base.yaml
├── values.yaml
└── release
    └── helmfile.yaml
```
With the following file contents
**helmfile.yaml**
```yaml
bases:
  - ../base.yaml
releases:
  - name: release
    chart: .
```
**base.yaml**
```yaml
environments:
  dev:
    values:
      - values.yaml
```
**values.yaml**
```yaml
foo: bar
```

I find that the path to `values.yaml` specified in `base.yaml` is relative to the `helmfile.yaml`, not the `base.yaml`.
```
$ ls
base.yaml  values.yaml  release

$ helmfile lint -e dev -f release/helmfile.yaml 
in release/helmfile.yaml: failed to read ../base.yaml: environment values file matching "values.yaml" does not exist in "."

$ cd release
$ helmfile lint -e dev
in ./helmfile.yaml: failed to read ../base.yaml: environment values file matching "values.yaml" does not exist in "."
```

I can fix this by changing my `base.yaml` to use the path `../values.yaml` instead. However, this solution doesn't work when I want to use a `base.yaml` fort multiple releases in different directory structures like below
```shell
.
├── base.yaml
├── values.yaml
├── release
│   └── helmfile.yaml
└── release2
    └── anotherdir
        └── helmfile.yaml
```

**Comment by Kellen275** [comment]
I found a similar issue from a while back: https://github.com/roboll/helmfile/issues/431
However, this applies to paths specified in the `chart: <path>` value of a release, which I'm not concerned with

**Comment by yxxhero** [comment]
@Kellen275 could you post a examle?

**Reply by Kellen275** [reply]
I believe the top of my original post is an example:
Given the following file structure
```shell
.
├── base.yaml
├── values.yaml
└── release
    └── helmfile.yaml
```
With the following file contents
**helmfile.yaml**
```yaml
bases:
  - ../base.yaml
releases:
  - name: release
    chart: .
```
**base.yaml**
```yaml
environments:
  dev:
    values:
      - values.yaml
```
**values.yaml**
```yaml
foo: bar
```

Running `helmfile lint -e dev` from the `release/` directory fails with the error
```shell
in ./helmfile.yaml: failed to read ../base.yaml: environment values file matching "values.yaml" does not exist in "."
```

**Reply by daveneeley** [reply]
I'm looking for the possibility to handle this this as well, the example is spot on for the problem.

**Comment by mumoshu** [comment]
Hey everyone!

TLDR;

- We need this, with some careful design.
- A workaround we might be able to use today is to use environment values. `base.yaml` has access to `.Environment.Values` of the consumer of `base.yaml`. So, you want pass arbitrary values from the consumer to the `base.yaml`, from the path to some directory base.yaml look values.yaml files for, or the values you want `base.yaml` to read.

> We need this, with some careful design

By design, each helmfile and base.yaml needs to be an independently runnable helmfile config to be tested separately. It avoids tight-coupling across multiple independent helmfile configs like helmfile.yaml and base.yaml.

If we changed `base.yaml` to refer to the `values.yaml` relative to its consumer's directory, it's a tight-coupling. `base.yaml` can be tested only with `values.yaml` that `base.yaml` does not know.

In the future, we may want to extend `bases` to support injecting values.yaml or the directory to look for extra values.yaml files.

It's also concerning that there is no easy way to let the consumer of `base.yaml` know about the necessity of providing `values.yaml` to the `base.yaml`. We don't have a "schema" for base.yaml today. Let's say you had an incorrect values file passed to `base.yaml`- `base.yaml` would just say `I have no value named blah-blah-blah`, without letting you know how to pass the value to it.

I'm eager to support this use case. It's just that I have not found a feasible way forward yet!
WDYT?

**Reply by daveneeley** [reply]
In the example, values are not passed to the base, but read from it.

An automatic property that resolves to the path of the current file would be sufficient for that case. If the property is not used the current behavior is preserved. Some scripting languages have this concept.

**Reply by daveneeley** [reply]
In Powershell there is an automatic variable called [MyInvocation](https://ss64.com/ps/syntax-myinvocation.html) .

> [$MyInvocation]...contains an object with information about the current command, such as a script, function, or script block.

> You can use the information in the object, such as the path and file name of the script ($MyInvocation.mycommand.path) or the name of a function ($myinvocation.mycommand.name) to identify the current command.

> The script below demonstrates the use of MyInvocation, n.b. this must be run as a script not interactively at the command line:
```
#[test.ps1]
" Display the full pathname of this running script:"
$myinvocation.mycommand.path
# This will return something like C:\demo\test.ps1
 
"`n Display the directory only (using a small function):"
 
function Get-ScriptDirectory {
  $Invocation = (Get-Variable MyInvocation -Scope 1).Value
  Split-Path $Invocation.MyCommand.Path
}

$dir = Get-ScriptDirectory
$dir
# This will return something like C:\demo
```

To restate the problem (mostly for my benefit), Helmfile works by changing the working directory to the base path of the currently executing helmfile, and all imports in bases that are imported by that helmfile are resolved relative to that working directory.

The workaround I chose was to put all helmfiles in the same directory (`helmfile.d`) so that any files imported in bases all have the same relative paths regardless of which helmfile they were imported by.

But taking this base file from the question:
```
environments:
  dev:
    values:
      - values.yaml
```
and changing it to something like this
```
environments:
  dev:
    values:
      - {{ clean (sprintf "%s/%s" (dir .MyInvocation.MyCommand.Path) "values.yaml") }}
```
Would get around the problem because `.MyInvocation.MyCommand.Path` would be the absolute path to the base file itself, and the location of the values.yaml would be resolved to an absolute path instead of a relative one. [sprig paths](https://masterminds.github.io/sprig/paths.html)
